tree-wide: nixos: /var/run -> /run

bobvanderlinden opened the pull request:
###### Motivation for this change

Fixes #51800 
Continuation of #47856 

`/var/run` is deprecated in `tmpfiles.d` and `/run` should be used. Currently this produces a warning every time `nixos-rebuild switch` runs for each service that uses `tmpfiles.d` with an entry for `/var/run`.

This PR changes `/var/run` instances into `/run` across nixpkgs.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

grahamc commented:
I'm not sure the individual commits with all the same commit message are usefully separate? May be good to squash them all together, or make their commit messages more specific about what they're doing.

worldofpeace commented on nixos/modules/system/boot/stage-2-init.sh:
```diff
@@ -85,15 +85,15 @@ install -m 0755 -d /var/{log,lib,db} /nix/var /etc/nixos/ \
 rm -rf /nix/var/nix/gcroots/tmp /nix/var/nix/temproots
 
 
-# For backwards compatibility, symlink /var/run to /run, and /var/lock
```
This comment could be left as is

worldofpeace commented on pkgs/tools/networking/network-manager/default.nix:
```diff
@@ -94,7 +94,7 @@ in stdenv.mkDerivation rec {
 
   preInstall = ''
-    installFlagsArray=( "sysconfdir=$out/etc" "localstatedir=$out/var" "runstatedir=$out/var/run" )
+    installFlagsArray=( "sysconfdir=$out/etc" "localstatedir=$out/var" "runstatedir=/run" )
```
This probably should be:
```suggestion
    installFlagsArray=( "sysconfdir=$out/etc" "localstatedir=$out/var" "runstatedir=$out/run" )
```

bobvanderlinden replied:
Keeping the old code in mind, your comment makes sense. However, I do wonder whether `runstatedir` should just be `/run` instead of `$out/run`.

worldofpeace replied:
Judging from the `installFlagsArray` it's so things get installed to $out/run and not /run which is outside the build prefix.

worldofpeace commented on nixos/modules/misc/locate.nix:
```diff
@@ -76,7 +76,7 @@ in {
     prunePaths = mkOption {
       type = listOf path;
-      default = ["/tmp" "/var/tmp" "/var/cache" "/var/lock" "/var/run" "/var/spool" "/nix/store"];
+      default = ["/tmp" "/run" "/var/tmp" "/var/cache" "/var/lock" "/var/run" "/var/spool" "/nix/store"];
```
This probably should be
```suggestion
      default = ["/tmp" "/var/tmp" "/var/cache" "/run/lock" "/run" "/var/spool" "/nix/store"];
```

bobvanderlinden replied:
Ah, `lock` is part of `run`. Looked it up, yes indeed.

bobvanderlinden commented on nixos/modules/programs/zsh/zsh.nix:
```diff
@@ -203,7 +203,7 @@ in
     environment.shells =
       [ "/run/current-system/sw/bin/zsh"
-        "/var/run/current-system/sw/bin/zsh"
+        "/run/current-system/sw/bin/zsh"
```
```suggestion
```

bobvanderlinden commented:
What needs to be done to get this merged? If doing this tree-wide is making the decision harder, I can split things up in separate PRs as well. For instance, I can do `/{var/,}run/kubernetes` separate from `/{var/,}run/current-system`

globin commented on pkgs/applications/display-managers/lightdm/run-dir.patch:
```diff
@@ -6,7 +6,7 @@ index 16b80f7..b3af435 100644
  #guest-account-script=guest-account
  #logind-check-graphical=false
  #log-directory=/var/log/lightdm
--#run-directory=/var/run/lightdm
```
That looks incorrect

globin commented on pkgs/applications/networking/remote/teamviewer/default.nix:
```diff
@@ -39,8 +39,6 @@ stdenv.mkDerivation rec {
     substituteInPlace $out/share/teamviewer/tv_bin/script/tvw_aux \
       --replace '/lib64/ld-linux-x86-64.so.2' '${glibc.out}/lib/ld-linux-x86-64.so.2'
-    substituteInPlace $out/share/teamviewer/tv_bin/script/tvw_config \
-      --replace '/var/run/' '/run/'
```
That should probably be kept

globin commented on pkgs/build-support/vm/default.nix:
```diff
@@ -111,7 +111,7 @@ rec {
     mkdir -p /fs/tmp /fs/run /fs/var
     mount -t tmpfs -o "mode=1777" none /fs/tmp
     mount -t tmpfs -o "mode=755" none /fs/run
-    ln -sfn /run /fs/var/run
```
Have you checked that this works correctly?

bobvanderlinden committed:
nixos/bash: do not use /var/run

bobvanderlinden committed:
nixos/fish: do not use /var/run

bobvanderlinden committed:
nixos/xonsh: do not use /var/run

bobvanderlinden committed:
nixos/zsh: do not use /var/run

bobvanderlinden committed:
nixos/bacula: /var/run -> /run

bobvanderlinden committed:
nixos/couchdb: /var/run -> /run

bobvanderlinden committed:
nixos/mongodb: /var/run -> /run

bobvanderlinden committed:
nixos/openldap: /var/run -> /run

bobvanderlinden committed:
nixos/redis: /var/run -> /run

bobvanderlinden committed:
nixos/rethinkdb: /var/run -> /run

bobvanderlinden committed:
nixos/pfix-srsd: /var/run -> /run

bobvanderlinden committed:
nixos/postgrey: /var/run -> /run

bobvanderlinden committed:
nixos/spamassassin: /var/run -> /run

bobvanderlinden committed:
nixos/matrix-synapse: /var/run -> /run

bobvanderlinden committed:
nixos/mbpfan: /var/run -> /run

bobvanderlinden committed:
nixos/spice-vdagentd: /var/run -> /run

bobvanderlinden committed:
nixos/svnserve: /var/run -> /run

bobvanderlinden committed:
nixos/nagios: /var/run -> /run

bobvanderlinden committed:
nixos/zabbix: /var/run -> /run

bobvanderlinden committed:
nixos/avahi-daemon: /var/run -> /run

bobvanderlinden committed:
nixos/bind: /var/run -> /run

bobvanderlinden committed:
nixos/hostapd: /var/run -> /run

bobvanderlinden committed:
nixos/htpdate: /var/run -> /run

bobvanderlinden committed:
nixos/ircd-hybrid: /var/run -> /run

bobvanderlinden committed:
nixos/lldpd: /var/run -> /run

bobvanderlinden committed:
nixos/miniupnpd: /var/run -> /run

bobvanderlinden committed:
nixos/ocserv: /var/run -> /run

bobvanderlinden committed:
nixos/raccoon: /var/run -> /run

bobvanderlinden committed:
nixos/supplicant: /var/run -> /run

bobvanderlinden committed:
nixos/xrdp: /var/run -> /run

bobvanderlinden committed:
nixos/fcron: /var/run -> /run

bobvanderlinden committed:
nixos/hologram-agent: /var/run -> /run

bobvanderlinden committed:
nixos/codimd: /var/run -> /run

bobvanderlinden committed:
nixos/restya-board: /var/run -> /run

bobvanderlinden committed:
nixos/selfoss: /var/run -> /run

bobvanderlinden committed:
nixos/tt-rss: /var/run -> /run

bobvanderlinden committed:
nixos/mighttpd2: /var/run -> /run

bobvanderlinden committed:
nixos/kodi: /var/run -> /run

bobvanderlinden committed:
nixos/xpra: /var/run -> /run

bobvanderlinden committed:
nixos/docker: /var/run -> /run

bobvanderlinden committed:
nixos/openvswitch: /var/run -> /run

bobvanderlinden committed:
nixos/cups: /var/run -> /run

bobvanderlinden committed:
nixos/tests: osquery: /var/run -> /run

bobvanderlinden committed:
nixos/tests: nghttpx: /var/run -> /run

bobvanderlinden commented:
Thanks a lot @globin for the review. I reverted the changes that you mentioned.  That made me realize there are a number of changes in here that are just wrong or edge cases. I went through the changes again, removed anything that seemed fishy and committed them separately.

Let me know whether this makes more sense.

Mic92 commented:
That needs to go to staging.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/solved-add-var-run-directory/1701/3

bobvanderlinden commented:
Rebased this and removed conflicting kubernetes commit, as the kubernetes module is now already using `/run`.

What can I do to move this forward?

oxij commented:
IMHO all the commit messages touching nixos subtree should be formatted as "nixos/<module>: ..." or similar (they should mention "nixos" at the very least). Also "tree-wide", not "tee-wide". Similarly, changes to pkgs should be formatted as "<package name>: ...".

Run

$ git filter-branch --msg-filter 'sed "1 { s%^module: %nixos/% ; s%^pkg: %% ; s%^test: %nixos/tests: % ; s%^tee-wide: %tree-wide: % ; }"' c18bb57627a87c6f03877386ecbc2e22d81d8afd..HEAD

and then fix the odd ones like "wpa_supplicant: ..." by hand.

As to the diffs, changes to nixos subtree LGTM. Changes to pkgs subtree are a mass rebuild so they should be split into a separate PR against staging.

bobvanderlinden committed:
nixos/asterisk: /var/run -> /run

bobvanderlinden committed:
nixos/wpa_supplicant: /var/run -> /run

bobvanderlinden commented:
@oxij Thanks a lot of the feedback! I've split the packages off to a separate PR: https://github.com/NixOS/nixpkgs/pull/58222

I split the wpa_supplicant and the asterisk commits in 2 (a nixos and a pkgs commit). The pkgs commit of those were also moved to #58222. It _should_ still be compatible if `/var/run` is indeed linked to `/run`

globin reviewed (approved):
Will merge in a few days if no other issues are brought up.

bobvanderlinden commented:
@globin merging would be awesome! :D
